common/src/main/java/io/netty/util/ConstantPool.java

```diff
@@ -68,7 +68,7 @@ public T valueOf(String name) {
     private T getOrCreate(String name) {
         T constant = constants.get(name);
         if (constant == null) {
-            final T tempConstant = newConstant(nextId(), name);
+            final T tempConstant = newConstant(0, name);
```

Comment:
always passing `0` here instead of `nextId()` seems like a semantic change. Although it is not clearly documented the `protected abstract T newContant(int id, String name);` interface appears to depend on some sort of "uniqueness" (as determined by `nextId()`) for `id`. Since `ConstantPool` is public there may be classes which extend it that are out of our control, and these classes may break as a result of this change.

Agreed we can't add a new abstract method, I've removed that from the code snippet above. However we also can't change the semantics of an existing method. So this means an alternative approach is to add a new method where we can define new semantics. I'm also open to alternative approaches that preserve the existing API semantics.